File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):

    answer = [0, 0]
    
    def is_all_same(arr, y, x, l):
        target = arr[y][x]    
        for j in range(y, y + l):
            for i in range(x, x + l):
                if arr[j][i] != target:
                    return False
        return True
    
    
    def press(arr, y, x, l):
        if is_all_same(arr, y, x, l):
            answer[arr[y][x]] += 1
            return
        
        
        press(arr, y, x, l // 2)
        press(arr, y + l // 2, x, l // 2)
        press(arr, y, x + l // 2, l // 2)
        press(arr, y + l // 2, x + l // 2, l // 2)
        
    
    press(arr, 0, 0, len(arr))
    
    return answer

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    int[] answer = {0, 0};
    
    public int[] solution(int[][] arr) {
        press(arr, 0, 0, arr.length);
        return answer;
    }
    
    public void press(int[][] arr, int y, int x, int l) {
        if (isAllSame(arr, y, x, l)) {
            answer[arr[y][x]]++;
            return;
        }
        
        press(arr, y, x, l / 2);
        press(arr, y + l / 2, x, l / 2);
        press(arr, y, x + l / 2, l / 2);
        press(arr, y + l / 2, x + l / 2, l / 2);
    }
    
    
    public boolean isAllSame(int[][] arr, int y, int x, int l) {
        int target = arr[y][x];
        
        for (int j = y; j < y + l; j++) {
            for (int i = x; i < x + l; i++) {
                if (target != arr[j][i]) {
                    return false;
                }
            }
        }
        return true;
    }
}